services/distribution/src/main/java/app/coronawarn/server/services/distribution/dgc/client/JsonSchemaDecoder.java

```diff
@@ -0,0 +1,118 @@
+    jsonSchemaMappingLookup = new JsonSchemaMappingLookup();
+  }
+
+  public JsonSchemaDecoder(ObjectFactory<HttpMessageConverters> messageConverters,
+      ObjectProvider<HttpMessageConverterCustomizer> customizers) {
+    super(messageConverters, customizers);
+  }
+
+  @Override
+  public Object decode(Response response, Type type) throws IOException, FeignException {
+    InputStream payloadJsonInputStream = response.body().asInputStream();
+    //String schemaPathToUse = getSchemaPathForReturnType(type);
+    String schemaPathToUse = getSchemaPathForRequestEndpoint(response.request().url());
+    if (schemaPathToUse == null) {
+      // no matching schema for this URL, so we don't need to validate
+      return super.decode(response, type);
+    }
+    InputStream schemaInputStream = resourceLoader.getResource(schemaPathToUse).getInputStream();
+    validateJsonAgainstSchema(payloadJsonInputStream, schemaInputStream);
+    return super.decode(response, type);
+  }
+
+  /**
+   * Resolve the schema that maps to the given feign client return type.
+   *
+   * @param requestUrl The endpoint of the request.
+   * @return The schema path for the given URL.
+   */
+  public String getSchemaPathForRequestEndpoint(String requestUrl) {
+    return jsonSchemaMappingLookup.getSchemaPath(requestUrl);
+  }
+
+  /**
+   * Validation logic to compare a json file to a schema.
+   *
+   * @param jsonPayloadInputStream The json payload.
+   * @param schemaAsStream The json schema.
+   * @throws IOException Thrown when json could not be loaded from file.
+   */
+  public void validateJsonAgainstSchema(InputStream jsonPayloadInputStream, InputStream schemaAsStream)
+      throws IOException {
+    try {
+      JSONObject jsonSchema = new JSONObject(new JSONTokener(schemaAsStream));
+      SchemaClient schemaClient = new ResourceSchemaClient(resourceLoader, JsonSchemaMappingLookup.JSON_SCHEMA_PATH);
+      Schema schema = SchemaLoader.load(jsonSchema, schemaClient);
+      //have to check manually if json is an array or an object. Limitation of the org.json library.
+
+      //workaround for the actual json in the stream -- if we give the InputStream to the JSONTokener directly below,
+      // it is saying that we don't have valid json (it days it should start with { or with [, which id DOES!)
+      String jsonPayloadString = new BufferedReader(
+          new InputStreamReader(jsonPayloadInputStream, StandardCharsets.UTF_8))
+          .lines()
+          .collect(Collectors.joining("\n"));
+
+      try {
+        JSONObject parsedObject = new JSONObject(new JSONTokener(jsonPayloadString));
+        schema.validate(parsedObject);
+      } catch (JSONException e) {
+        try {
+          JSONArray parsedArray = new JSONArray(new JSONTokener(jsonPayloadString));
+          parsedArray.forEach(object -> schema.validate(object));
+        } catch (JSONException ne) {
+          throw new RuntimeException("json is neither an object nor an array");
```

Comment:
```suggestion
          throw new RuntimeException("json is neither an object nor an array", ne);
```

`JSONException` itself is already a RuntimeException... theoretically it's not necessary at all to catch it...